### DISCUSSION TITLE
关于键盘UI自定义的想法

### DISCUSSION THREAD

**Discussion by imfuxiao** [imfuxiao/Hamster#172] [Announcements]
键盘UI自定义马上进入开发阶段，大家有什么想法可以在这里讨论。

**Comment by imfuxiao** [comment]
原则1: 将目前滑动展示内容从滑动设置中移除，展示功能放在UI设置中。

**Reply by imfuxiao** [reply]
原则2: 键盘上面的字母可以自由设置大小写。#47

**Reply by imfuxiao** [reply]
原则3: 切换英文模式能有不同的键位，且英文键盘的上滑功能与中文键盘可以独立分开。

**Reply by 23fo** [reply]
> 原则2: 键盘上面的字母可以自由设置大小写。https://github.com/imfuxiao/Hamster/issues/47

这个是不是可以采用下面这个方案https://github.com/imfuxiao/Hamster/discussions/172#discussioncomment-5884082

> 按鍵中央顯示名稱應該採用translator/preedit_format（尤其是定義了xlit映射的情況下）。

可以设置大小写，想设什么设成什么

**Comment by doraeba** [comment]
1、键盘面板的高度，给用户自定义权限；
2、九宫格按钮大小和间隙,希望可以借鉴下系统输入法；
3、剪贴板，或者叫便签功能(刚需，参考下落格的快捷短语)，希望大佬能增添上，并给予点划唤醒支持。

**Reply by imfuxiao** [reply]
第三点后面描述的点划唤醒没理解是什么功能，你提的第三点非常不错，另开个主题吧，这个主要讨论做键盘UI的功能。

**Reply by doraeba** [reply]
上/下滑某个键位(需用户预设)，唤出剪贴板(快捷便签)面板😄

**Reply by imfuxiao** [reply]
了解~

**Reply by xlztong** [reply]
想问问落格的快捷短语是什么样

**Reply by doraeba** [reply]
![image](https://user-images.githubusercontent.com/52118447/234891171-9057c0b3-f413-4b90-8fe4-abe09d570cb3.png)

![image](https://user-images.githubusercontent.com/52118447/234891219-9651527c-568b-492a-9276-2e1eb8ee3e13.png)

**Comment by wzxmer** [comment]
1.按键之间上下左右间隔也能自定义（等于是自定义按键大小了）
2.能定义键盘背景图
3.能对单个按键进行UI设定，比如边框的颜色，背景的颜色，字体的颜色，字体的大小

**Reply by imfuxiao** [reply]
关于背景图可以具体展开说下嘛

**Reply by wzxmer** [reply]
大概是：类似于整个键盘透明化（以及键盘上的按键字体也能自定义清除，只留空白），可以选择本地图片（icloud）上传设置成背景，这点可以参考搜狗

**Reply by wzxmer** [reply]
> 关于背景图可以具体展开说下嘛

找人问了一下。
1.本地化 替换图片
2.提供svg 参数变更支持（主体svg 需提供上传窗口）

**Comment by androidcn** [comment]
![image](https://user-images.githubusercontent.com/271715/234885558-dcd65e7a-b23a-42a5-b551-ca2fda60a65f.jpeg)

习惯百度输入法的布局和手感

**Comment by Nicked639** [comment]
- 增加剪贴板上屏功能（也许不能算UI😅）
- 默认键盘可增加不显示滑字符的开关（滑动后再予以显示）
- 点选数字键123后弹出的标点符号键盘提供自定义替换标点功能

**Comment by hugh7007** [comment]
希望可以给大佬一定的思路，但自己实在小白，希望不会对大佬的工作造成困扰。
一个按钮的组成  分为  宽+高+弧度+颜色  + 文字 
需要操控时，点击对应按钮可实现修改对应按钮的参数 

一个键盘的间距是固定的。但给多大编辑范围才是需要好好考虑，目前世面上也没有绝对正确的选项。所以尝试了一个思路。

需要控制键盘，点击键盘设置 →  出现虚拟（原始键盘）   →   上面分别依次为  按钮横间距  列间距  背景颜色     下图为一个虚拟键盘  ，点击哪个修改哪个 

下方

<img width="1506" alt="微信图片_20230428085807" src="https://user-images.githubusercontent.com/48746344/235029869-81642cab-2676-45ef-893e-f6d45ac36e6a.png">

**Reply by imfuxiao** [reply]
收到，感谢您为此做了UI设计，您的UI设计对我非常有帮助。同时您的建议也很棒。

目前来说键盘自定义布局还没有一个完整的思路，此功能又非常重要，所以想听听大家的意见与建议。

等我有了完整思路，我会提出提案，到时希望您也能参与。

**Reply by hugh7007** [reply]
ok,没有问题，同时关于
1.贴图部分 ，这个思路是没问题的。  < 如何解决会造成过大的问题，原始底图设置足够的情况下，其实是可以考虑的 ><缺点，ipad显示问题，虽然可以通过读取设备来规避一些问题，但并不能保证软件不出错>
2.svg部分，我也可以提供部分的支持。  <目前世面上最常见的做法，但svg也是有两种分支，一种是单按键 ， 一种是整体键盘。> <缺点：繁杂的设置><用户需要自己制作时的门槛>
如果我没有及时回复的话<因为实在非经常使用这个网站，目前是专门为了支持大佬来的>
通过联系：wzxmer  ，可随时转达给我。或者我在tg里详细与您对话

**Comment by LeoCao-X** [comment]
请问后续会支持微软双拼定制的UI吗，目前如果使用微软双拼，由于缺乏键盘上面的; 会导致输入法无法输入该键位代表的音

**Comment by AaronConlon** [comment]
请问有支持 9 宫格输入 UI 的计划吗？目前看来四叶草简体拼音没办法用九宫格输入 感谢回复~ @imfuxiao

**Comment by THMonster** [comment]
不知道这个和浮动键盘是否有关。在ipad下目前在浮动键盘下该输入法的界面整个是错乱的，似乎是还没有适配？

**Reply by imfuxiao** [reply]
抱歉，应该是的，ipad没有好好测试，回头我看下。

**Comment by ufo169** [comment]
rime的魅力，在于高度自定义，简单的固定键盘布局，会大大降低它对方案的适应性，比如九宫格输入法没有拼音筛选，需要额外的添加笔画筛选，或者自定义的三拼方案等，很多顶功方案也需要添加额外的笔划按键，还有很多方案还需要数字按键的辅助。
对于按键元素，最基本的自定义输出，显示符号，比如说，现在的九宫数字键盘，好像为了不常用的计算器，布局没有空格也没有@.够痛苦的，
为了降低误触，按键的大小，间隙，圆角，显示符号和滑动输出的符号的大小，偏移，还有少量的阴影（稍微的立体感避免误触有很大帮助），按键的颜色自定义，功能按键和特殊位置的颜色加持，用起来感觉完全不同。对于图片功能，高消耗性能慎重处理，同文也吃过这样的亏
按键反馈，震动，声音的自定义感觉也很有必要，

**Reply by imfuxiao** [reply]
感谢，建议非常有用。

**Comment by LoyaltyMu** [comment]
1.同楼上增加剪贴板功能。
2.字母可以支持左右滑动打出大写或小写英文字母

**Reply by wzxmer** [reply]
> 1.同楼上增加剪贴板功能。 2.字母可以支持左右滑动打出大写或小写英文字母

左右滑，默认已有支持，默认生成的配置文件不带设置左右，需要手动导出修改文件进行添加，这个会进行优化的

**Comment by DawnLiExp** [comment]
建议使用外部解决方案：https://github.com/LEOYoon-Tsaw/Squirrel-Designer
生成代码自己编辑文件写入，这样就不会增加app的负载了。

**Reply by imfuxiao** [reply]
这个想法非常棒~

**Comment by groverlynn** [comment]
1. 按鍵中央顯示名稱應該採用`translator/preedit_format`（尤其是定義了`xlit`映射的情況下）。
2. 提供四行鍵盤（`1QAZ`）選項，甚至應該作爲默認佈局，因爲不少輸入法是需要使用數字行輸入的。
3. 增加switch呼出按鈕（對應默認的`Ctrl+｀`）

**Comment by KyleBing** [comment]
建议主键盘区跟苹果官方的五笔输入法键位保持一致，至少每个键的位置保持完全一致。这样用起来会舒服些。

**Comment by keaidu** [comment]
长按键盘可以输入符号或大小写或符号，英语输入可以关联单词。还有空格滑动光标，这个真的很好用。
![Uploading IMG_6578.png…]

**Reply by keaidu** [reply]
![Uploading IMG_6578.png…]()

**Comment by chuvily** [comment]
请问,可以增加14键的方案吗?这个按键大小实在是hold不住啊

**Reply by imfuxiao** [reply]
嗯嗯，需要等自定义键盘功能上线后，自己就可以布局了。

**Reply by chuvily** [reply]
对了，希望可以增加助记功能。。无论是trime一样利用配制文件实现也好，还是利用背景图片实现也罢。这个功能说实话，无论是五笔或是双拼，或是音形都是非常使用的功能。

**Comment by Anber55** [comment]
希望能够增加

1️⃣字母、汉字、上下滑的内容的颜色深度跳转

2️⃣能够键盘增加一行，自由调整空隙
3️⃣表情页和符号页都能在上下滑功能中实现
4️⃣所有键支持长按操作
5️⃣可以支持所有键的实际字母和显示内容，例如实际的是shift，显示的为🪴

**Comment by 4tiger** [comment]
1. 非常感谢作者，尝试了不少输入法，目前iOS下面最好的。很希望有个链接，请作者喝杯咖啡。
2. 键盘上是否可以调整字体大小？字母都还好了，按键滑动手势对应的字符实在太小，很难看的清楚。

**Comment by amorphobia** [comment]
一个不成熟的设想：用 yaml 语法映射到 SwiftUI 的一个子集

```yaml
main_keyboard:
  type: vstack
  rows:
    - name: row1
      type: hstack
      height: 20
      buttons:
        - { name: q, click: q, width: 10 }
        - { name: w, click: w, width: 10 }
        - ...
    - name: row2
      height: 20
      bottons:
        - ...
    - ...
```

我不太懂 SwiftUI, 这里只是我简单的设想，真的实现的时候可能还要考虑有的按钮是可以跨行的，还有就是类似现在九宫格数字键盘里可以滑动的符号列表等等……

**Reply by imfuxiao** [reply]
嗯嗯，跨行跨列的确实不好处理。目前数字九宫格的纯手工布局还好处理，真的要自定义实现就难了。不清楚别的输入法自定义键盘是否支持跨行跨列的按键。

**Reply by Jmdayou** [reply]
有没有可能把键盘区域做一个画布（可手动自定义大小）
通过左上坐标定义按键位置 利用长宽属性定义大小
最好做成可视化进行拖动
可以预先设定几个功能（比如空格、删除、shift、键盘切换等），自定义按键的键值
同时按键支持分配按下、长按、各个方向滑动的键值，还有颜色、显示的标签、圆角大小等等属性。
配置完成可自动生成配置代码

感觉有点难😂

**Reply by imfuxiao** [reply]
这几天在做前期工作，慢慢来吧。

**Reply by ufo169** [reply]
就实用而言，可以滑动的符号列表布局，还不如平分来的方便快捷。因为滑动输入比较慢，而且，又设置不了各方向的滑动输入。比如：
![image](https://github.com/imfuxiao/Hamster/assets/78493529/c1675764-218c-48cf-ac12-2b1346e11d5f)
对于跨行和平分，设置起来确实需要点技巧，需要设置一些占位空按键。就兼容性而言，还是习惯了yaml格式的自定义。

**Comment by keaidu** [comment]
能不能增加一个类似搜狗输入法的英文输入模式啊？

ufo169 ***@***.***>于2023年6月8日 周四22:47写道：

**Reply by ufo169** [reply]
仓，有一个切换方案功能（比如，上滑123摁键），你设置一个中文方案和一个英文输入方案，一上滑123摁键，就切换到英文输入方案，再切换，又回到中文输入方案。

**Comment by keaidu** [comment]
英文输入没办法联想单词，以后会添加吗？

ufo169 ***@***.***>于2023年6月11日 周日16:24写道：

**Reply by ufo169** [reply]
你去GitHub搜一个，英文输入方案，要啥功能没有，英文效果如下
![image](https://github.com/imfuxiao/Hamster/assets/78493529/6c7415c6-f84b-404f-81d7-04008c14f6f2)

**Reply by keaidu** [reply]
这个输入方案我试了，然后装到Rime文件配置完没法用，方案栏里找不到。

**Reply by ufo169** [reply]
那是，你连续输入增强功能没配置好，简单的可以，把lua功能关闭，#    - lua_filter@easy_en_enhance_filter

**Reply by keaidu** [reply]
配置完直接把那几个文件丢Rime里就行了么？

**Reply by ufo169** [reply]
如果注释了lua功能，那简单的把方案和词典文件扔进去，添加方案，再配置一下切换键重新部署就行，

**Reply by Nemophilister** [reply]
> 你去GitHub搜一个，英文输入方案，要啥功能没有，英文效果如下
> 
> ![image](https://github.com/imfuxiao/Hamster/assets/78493529/6c7415c6-f84b-404f-81d7-04008c14f6f2)
> 
> 
> 
> 
> 
> 

大佬求教我去找了没有找到英文输入法

**Reply by Nemophilister** [reply]
> 你去GitHub搜一个，英文输入方案，要啥功能没有，英文效果如下
> 
> ![image](https://github.com/imfuxiao/Hamster/assets/78493529/6c7415c6-f84b-404f-81d7-04008c14f6f2)
> 
> 
> 
> 
> 
> 

![image](https://github.com/imfuxiao/Hamster/assets/136904652/78dad01b-f1f6-4e52-a029-f645d3638b5c)找到就是你这个

**Comment by keaidu** [comment]
感谢

ufo169 ***@***.***>于2023年6月11日 周日20:59写道：

**Comment by kanwakyuudai** [comment]
這是我使用的自定義布局，謹供參考：
**1.全拼布局：**
![Image (1)](https://github.com/imfuxiao/Hamster/assets/17743934/e2e4b1ec-96ae-438e-9db0-4c6851d52564)
v替換成拼音的ü
**2.注音布局：**
![Image](https://github.com/imfuxiao/Hamster/assets/17743934/556804dc-ce7c-44f8-8e2c-7caeaf7f0528)
聲調符非必須，可以替換成標點
**3.九宮格小鍵盤布局：**
![Image 2](https://github.com/imfuxiao/Hamster/assets/17743934/647d3b0d-d01f-4a84-a7e5-4ffdbd597736)
類似計算器布局
**4.英文布局：**
![Image 1](https://github.com/imfuxiao/Hamster/assets/17743934/012d838c-55fa-4ac8-bbe2-fdf5ea49381e)
只輸出半角字符

**Comment by evercyan** [comment]
用的五笔, 然后手机上全键盘误触多, 所以一直用的这种双键布局, 辛苦大佬看下, 

![image](https://github.com/imfuxiao/Hamster/assets/10394477/5b5264b9-4e8f-47d5-92ff-2910ac11cbbc)

**Comment by er1c-zh** [comment]
定义键盘的布局也许可以参考这种格式？ http://www.keyboard-layout-editor.com/

**Reply by imfuxiao** [reply]
> 定义键盘的布局也许可以参考这种格式？ http://www.keyboard-layout-editor.com/

网站不错，感谢。

**Comment by LoyaltyMu** [comment]
仓有办法支持双拼滑动输入吗

**Reply by imfuxiao** [reply]
抱歉，目前没有滑动输入功能。

**Comment by hypan599** [comment]
请问每个键显示的字母可以用`clover.custom.yaml`定义吗? 我改`clover.custom.yaml`里面`  key_binder.bindings`可以实现Colemak的布局,但是就变成了和UI字母不一致,非常别扭. (比如说按 r 会打出 s, Colemak里面那里就是s).

如果好做的话,希望可以简单加一个开关让每个键不显示字母, 只是一块空白的键盘,作为Colemak等等用户 workaround. (毕竟都用Colemak了,键位背下来都好说,只要别有misleading就好)

**Reply by imfuxiao** [reply]
自定义键盘功能还未开发完毕，目前只有标准QWERT键盘布局。后续自定义键盘功能上线后，按照此功能设置是可以支持的。

**Reply by harryhook** [reply]
期待早日发布

**Reply by imfuxiao** [reply]
> 期待早日发布

嗯嗯，我尽量。本次功能对我来说是个不小的挑战，所以会慢很多，还有很多技术债要补，另外还对旧版本代码做了重构。

总之，我尽量快些。

**Comment by tywtyw2002** [comment]
建议键盘布局可以通过Yaml或者其他类似的config类文件进行自定义按键。

这样可以充分的自定义布局，有点像qmk/tmk之类的机械键盘固件，按键你想定义什么keycode就是什么keycode。

自定义键盘配置其实可以进行设置，这个按键是直接ascii输出，还是走librime，这样可以解决一些中英文符号问题。


布局可以进行预设，直列模式（九宫格）还是，斜列（正常qwert），按键的宽度，1u，1.25u。。。。。这样code好写，又可以最大化的自定义键盘布局。

**Comment by Jreen** [comment]
9宫格那个大写英文我个人觉得可能有点大，导致有紧密感。所以不知道能不能支持修改比如说abc的文字大小以及abc之间的紧密感？

**Reply by imfuxiao** [reply]
抱歉，目前还不能自己调节。后面我会优化下，感谢反馈。

**Comment by cj233cj** [comment]
请问2.0更新后要把仓内置的自然码双拼方案键盘修改为类似原生单手键盘的宽度应该具体修改哪个文件

**Reply by imfuxiao** [reply]
您好，单手模式功能还未开发。
